wallet: Fix relative path backup during migration.

davidgumberg opened the pull request:
Support for wallets outside of the default wallet directory was added in #11687, and these external wallets can be specified with paths relative to the wallet directory, e.g.  `bitcoin-cli loadwallet ../../mywallet`. In the RPC commands, there is no distinction between a wallet's 'name' and a wallet's 'path'. This PR fixes an issue with wallet backup during migration where the wallet's 'name-path' is used in the backup filename. This goes south when that filename is appended to the directory where we want to put the file and the wallet's 'name' actually gets treated as a path:

```cpp
    fs::path backup_filename = fs::PathFromString(strprintf("%s_%d.legacy.bak", (wallet_name.empty() ? "default_wallet" : wallet_name), GetTime()));
    fs::path backup_path = this_wallet_dir / backup_filename;
```

Attempting to migrate a wallet with the 'name' `../../../mywallet` results in a backup being placed in `datadir/wallets/../../../mywallet/../../../mywallet_1744683963.legacy.bak`.

If permissions don't exist to write to that folder, migration can fail.

The solution implemented here is to put backup files in the top-level of the node's `walletdir` directory, using the folder name (and in some rare cases the file name)  of the wallet to name the backup file:

https://github.com/bitcoin/bitcoin/blob/9fa5480fc4c46fa11345c45fbe4dd21c127e3e05/src/wallet/wallet.cpp#L4254-L4268


##### Steps to reproduce on master
Build and run `bitcoind` with legacy wallet creation enabled:
```bash
$ cmake -B build -DWITH_BDB=ON && cmake --build build -j $(nproc)
$ ./build/bin/bitcoind -regtest -deprecatedrpc=create_bdb
```

Create a wallet with some relative path specifiers (exercise caution with where this file may be written)

```bash
$ ./build/bin/bitcoin-cli -regtest -named createwallet wallet_name="../../../myrelativewallet" descriptors=false
```

Try to migrate the wallet:
```bash
$ ./build/bin/bitcoin-cli -regtest -named migratewallet wallet_name="../../../myrelativewallet"
```

You will see a message in `debug.log` about trying to backup a file somewhere like: `/home/user/.bitcoin/regtest/wallets/../../../myrelativewallet/../../../myrelativewallet_1744686627.legacy.bak` and migration might fail because `bitcoind` doesn't have permissions to write the backup file.

davidgumberg committed:
wallet: Fix migration of wallets with pathnames.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

davidgumberg committed:
test: Migration of a wallet with `../` in path.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4500,9 +4500,11 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
-    fs::path backup_filename = fs::PathFromString(strprintf("%s_%d.legacy.bak", (wallet_name.empty() ? "default_wallet" : wallet_name), GetTime()));
-    fs::path backup_path = this_wallet_dir / backup_filename;
+    fs::path legacy_wallet_dir = fs::absolute(fs::PathFromString(local_wallet->GetDatabase().Filename())).parent_path();
+    // Use the wallet directory's name, rather than the wallet name, to avoid issues with wallet's whose names contain relative file paths.
```
In commit "wallet: Fix relative path backup during migration." (db3974ee11b50c5cdccd277aa6d5cce660032718)

This comment doesn't seem right. A relative file path like "foo" would not cause issues here. The problems happen with paths that contains slashes, which is why the line below is calling `filename()` to drop everything before the last slash. 

Would suggest a different comment that explains what this is trying to do. Maybe: "Create a backup copy of wallet data inside the wallet directory with a unique and descriptive filename, based on the name of the wallet and the current timestamp. If this wallet is not the default wallet, the last component of the wallet's directory name is used as the backup filename since the full wallet name may contain slashes."

davidgumberg replied:
> This comment doesn't seem right.

This was wrong, thanks! I've added a variation of the comment you suggested that I edited for brevity.

ryanofsky reviewed (approved):
Code review ACK 1daaeedfafa0ee480a10756fa2a5cb25a268a3fa. Nice find and fix!

> I suspect this may not be the only bug like this, since everywhere that a wallet's name appears in code, that name might be a path

Wallet paths are meant to be passed to the database driver once during opening and not manipulated or interpreted after that. Backup and migration code are an exception because they do interpret the paths after they are opened, so that code needs to be written in a consistent way, but it should be a small amount of code.

ryanofsky reviewed (approved):
Code review ACK 7343538aa35101815f2823d1f102bc990abc1232, just changing a comment since the review

re: https://github.com/bitcoin/bitcoin/pull/32273#issue-2994969229

> This does not cause an issue for wallets with absolute pathnames 

IIUC, this is mostly true, but not totally. If intention for previous code was to place the wallet backup file inside the wallet directory, it looks like that was not previously happening, and that was also a bug which is now fixed.

Also IIUC, there wasn't an issue with relative paths generally like "a/b/c", only an issue with relative paths containing `..` pointing outside of the wallets directory? Might be good idea to change title to something like "Fix external wallet backups during migration". But title also seems ok currently since the detailed description is pretty clear.

achow101 commented:
Can you add a test for for failing to migrate a wallet specified by relative path? It's possible that restoration won't work as we expect it to.

davidgumberg committed:
test: Migration fail recovery w/ `../` in path

davidgumberg commented:
> Can you add a test for for failing to migrate a wallet specified by relative path? It's possible that restoration won't work as we expect it to.

Added a test that the 'automatic' restoration from a backup after a failed migration works, and added a test that restoring a backup made from an external relative wallet works.

> Also IIUC, there wasn't an issue with relative paths generally like "a/b/c", only an issue with relative paths containing `..` pointing outside of the wallets directory?

I thought so too, but checked after reading your comment and this is also broken on master:

```bash
./build/bin/bitcoin-cli -regtest -named createwallet wallet_name="a/b/c/d" descriptors=false
./build/bin/bitcoin-cli -regtest -named migratewallet wallet_name="a/b/c/d"
```

results in:

```console
error copying /home/user/.bitcoin/regtest/wallets/a/b/c/d/wallet.dat to
/home/user/.bitcoin/regtest/wallets/a/b/c/d/a/b/c/d_1744848074.legacy.bak
```

pseudoramdom commented:
Tested ACK https://github.com/bitcoin/bitcoin/pull/32273/commits/db4ea2632f0c1a5a5d6885526545cbddae7305d7

The patch works for both  paths containing `"../"` and `"a/b/c/d"`
> I thought so too, but checked after reading your comment and this is also broken on master:

+1. `master` does fail `migrate` for `"a/b/c"` style.

<details>
<summary>master</summary>

Migrating with `..` style
```
2025-04-17T01:49:47Z copied /Library/Application Support/Bitcoin/regtest/wallets/../../../myrelativewallet/wallet.dat to /Library/Application Support/Bitcoin/regtest/wallets/../../../myrelativewallet/../../../amyrelativewallet_1744854587.legacy.bak
```

Migrating with `a/b/c` style
```
2025-04-17T02:12:59Z error copying /Library/Application Support/Bitcoin/regtest/wallets/a/b/c/myrelativewallet/wallet.dat to /Library/Application Support/Bitcoin/regtest/wallets/a/b/c/myrelativewallet/a/b/c/myrelativewallet_1744855979.legacy.bak - filesystem error: in copy_file: No such file or directory ["/Library/Application Support/Bitcoin/regtest/wallets/a/b/c/myrelativewallet/a/b/c/myrelativewallet_1744855979.legacy.bak"] ["/Library/Application Support/Bitcoin/regtest/wallets/a/b/c/myrelativewallet/wallet.dat"]
```
</details>

<details>
<summary>PR 32273</summary>

Migrating with `..` style
```
2025-04-17T01:58:38Z copied /Library/Application Support/Bitcoin/regtest/wallets/../../../myrelativewallet2/wallet.dat to /Library/Application Support/Bitcoin/regtest/wallets/../../../myrelativewallet2/myrelativewallet2_1744855118.legacy.bak
```

Migrating with `a/b/c` style
```
2025-04-17T02:15:39Z copied /Library/Application Support/Bitcoin/regtest/wallets/a/b/c/myrelativewallet2/wallet.dat to /Library/Application Support/Bitcoin/regtest/wallets/a/b/c/myrelativewallet2/myrelativewallet2_1744856139.legacy.bak
```
</details>

davidgumberg commented on test/functional/wallet_migration.py:
```diff
@@ -925,6 +990,54 @@ def test_failed_migration_cleanup(self):
+        assert_equal(original_shasum, new_shasum)
+
+        # Check the wallet we tried to migrate is still BDB
+        datfile = os.path.join(absolute_path, "wallet.dat")
```
Writing as a review comment since I didn't think a code comment was warranted here, the reason for putting this on a separate line even though it is only used once is to avoid the wrath of the slightly [imprecise](https://github.com/bitcoin/bitcoin/blob/e66e30c9e5383a467789574e61114b57536193b9/test/lint/lint-python-utf8-encoding.py#L31) python-utf8-encoding [lint check](https://github.com/bitcoin/bitcoin/blob/e66e30c9e5383a467789574e61114b57536193b9/test/lint/lint-python-utf8-encoding.py) which otherwise won't catch that the file being opened in binary mode.

shahsb commented on test/functional/wallet_migration.py:
```diff
@@ -531,6 +532,70 @@ def test_unloaded_by_path(self):
+        relative_name = os.path.relpath(absolute_path, start=self.master_node.wallets_path)
+
+        wallet = self.create_legacy_wallet(relative_name)
+        # listwalletdirs only returns wallets in the wallet directory
```
You could reframe the comment to explain why this is significant:

```
# Wallet created using relative path won't reside in the default wallet dir,
# hence it should not appear in listwalletdir results.
```

davidgumberg replied:
Rather than explaining the line below, I added this comment to add context that I think isn't immediately available from the test, namely what the (obscure in both use and name) `listwalletdir` rpc does, I think your suggested comment would repeat what the test is asserting.

shahsb commented on test/functional/wallet_migration.py:
```diff
@@ -531,6 +532,70 @@ def test_unloaded_by_path(self):
+        common_parent = os.path.commonpath([self.master_node.wallets_path, self.old_node.wallets_path])
+
+        # This test assumes that the relative path from each wallet directory to the common path is identical.
+        assert_equal(os.path.relpath(common_parent, start=self.master_node.wallets_path), os.path.relpath(common_parent, start=self.old_node.wallets_path))
```
If this assumption ever fails (say due to environment setup differences), the test will break early.
Consider failing gracefully with a more informative message:

```
assert_equal(..., ..., "Relative paths from wallet dirs to common parent must be identical for this test.")
```

davidgumberg replied:
Unlike, [`assert_not_equal`](https://github.com/bitcoin/bitcoin/blob/6b4b7ac2113ed969bd31ce42190b96bd2e1b9c97/test/functional/test_framework/util.py#L79-L81), `assert_equal` [does not support](https://github.com/bitcoin/bitcoin/blob/6b4b7ac2113ed969bd31ce42190b96bd2e1b9c97/test/functional/test_framework/util.py#L72-L77) passing an error message. Re: the conceptual change; while it might have some advantages over the style currently used, as far as I can tell it is almost never the case that we log the "big picture" of what went wrong in a functional test, rather than details about the exact comparison or check that failed.

shahsb commented on test/functional/wallet_migration.py:
```diff
@@ -531,6 +532,70 @@ def test_unloaded_by_path(self):
+        # Check that old node can restore from the backup.
+        self.old_node.restorewallet("relative_restored", migrate_res['backup_path'])
+        wallet = self.old_node.get_wallet_rpc("relative_restored")
+        assert wallet.gettransaction(txid)
```
Can we add more stricter checks on the transaction received here.

For example:
```
tx = wallet.gettransaction(txid)
assert_equal(tx["amount"], 1)
assert_equal(tx["confirmations"], 1)
```
This gives you tighter guarantees and helps with debugging if the test fails.

davidgumberg replied:
We are already asserting the transaction id exists in the migrated wallet, and that the wallet has the correct balance, I'm not totally opposed to checking the tx amount as opposed to the wallet balance, but I'm not sure that this change really does add any guarantees, or help with debugging when the test fails, so I will err on the side of keeping this as-is so it [parallels](https://github.com/bitcoin/bitcoin/blob/9a4c92eb9ac29204df3d826f5ae526ab16b8ad65/test/functional/wallet_migration.py#L530-L532) `test_unloaded_by_path` above, although after testing, I realized that the `assert` I had here does not do anything, since if `gettransaction` fails, it results in a `JSONRPCException`.

shahsb commented on test/functional/wallet_migration.py:
```diff
@@ -925,6 +990,54 @@ def test_failed_migration_cleanup(self):
+
+        assert all(wallet not in self.master_node.listwallets() for wallet in [f"{wallet_name}", f"{wallet_name}_watchonly", f"{wallet_name}_solvables"])
+
+        assert not (self.master_node.wallets_path / "failed_watchonly").exists()
```
Nit: May be a small comment with reason would help here:
```
# No 'failed_watchonly' directory should exist because migration cleanup should be complete
```

shahsb commented on test/functional/wallet_migration.py:
```diff
@@ -925,6 +990,54 @@ def test_failed_migration_cleanup(self):
+        common_parent = os.path.commonpath([self.master_node.wallets_path, self.old_node.wallets_path])
+
+        # This test assumes that the relative path from each wallet directory to the common path is identical.
+        assert_equal(os.path.relpath(common_parent, start=self.master_node.wallets_path), os.path.relpath(common_parent, start=self.old_node.wallets_path))
```
Similar to previous test and all other places, can we fail gracefully with more informative message:
```
assert_equal(..., ..., "Test requires that both wallet paths have the same relative path to the common parent")
```

davidgumberg replied:
See https://github.com/bitcoin/bitcoin/pull/32273#discussion_r2056813556.

shahsb commented on test/functional/wallet_migration.py:
```diff
@@ -925,6 +990,54 @@ def test_failed_migration_cleanup(self):
+
+        # Check the wallet we tried to migrate is still BDB
+        datfile = os.path.join(absolute_path, "wallet.dat")
+        with open(datfile, "rb") as f:
```
We should first validate that file is present at the specified location and accessible and then only try to open and validate other things

davidgumberg replied:
If `open` fails an exception will be thrown and the block below it will not be executed.

https://docs.python.org/3/reference/compound_stmts.html#with

shahsb commented on test/functional/wallet_migration.py:
```diff
@@ -925,6 +990,54 @@ def test_failed_migration_cleanup(self):
+
+        # Check the wallet we tried to migrate is still BDB
+        datfile = os.path.join(absolute_path, "wallet.dat")
+        with open(datfile, "rb") as f:
```
**Minor / Nit:** A separate helper function could be introduced to validate it's BDB for better readability.
Example:
```
def is_bdb_wallet(filepath):
    with open(filepath, "rb") as f:
        data = f.read(16)
        _, _, magic = struct.unpack("QII", data)
        return magic == BTREE_MAGIC
```

and then:
```
assert is_bdb_wallet(datfile)
```

davidgumberg replied:
I'll leave this as-is for now.

shahsb commented:
Concept ACK

ryanofsky reviewed (approved):
Code review ACK 6b4b7ac2113ed969bd31ce42190b96bd2e1b9c97. Main change is expansion of test to cover restoring from the backup.

IIUC, PR description seems helpful for describing part of the problem, but contains inaccuracies and would be good to revise to avoid being misleading. Here are things I might add or correct:

1. PR describes there being a problems with relative paths, but not all relative paths have a problem, and absolute paths have problems too. Things only currently work correctly when wallet names contains no slashes (e.g. `mywallet`). Whenever a wallet name has slashes, there are bugs which this PR fixes (the bugs are different depending on whether the path is absolute or relative).
2. If wallet name is relative and has slashes like `my/wallet` there is a bug currently because migrate code tries to write a backup file inside the wallet directory at a path like `my/wallet/my/wallet_123.bak` and this fails when the backup file can't be written at a nonexistent subdirectory. This PR fixes that problem by dropping extra path components so the backup is saved at `my/wallet/wallet_123.bak`
3. If wallet name is absolute and has slashes like `/my/wallet` there is a bug currently because instead of the backup being written inside the wallet directory at a location like `/my/wallet/wallet_123.bak` the backup is written outside the wallet directory, next to it,  at a location like `/my/wallet_123.bak`. The PR fixes that problem by always writing the backup file inside the wallet directory.
4. If wallet name is relative and contains `..` there can be a combination of both problems: an attempt to write the backup file outside the wallet directory can happen, and the write can fail because the destination directory probably won't exist.

Also, I think I'd suggest dropping "I suspect this may not be the only bug like this, since everywhere that a wallet's name appears in code, that name might be a path, and that path may contain relative specifiers, and the consequences of that are not always obvious, and test coverage for wallets with relative paths seems like it could be improved." because it increases vagueness and uncertainty in the PR description and I don't actually think it is true. Outside of backup and restore code wallet names are treated like identifiers, not like paths, so I don't think we should expect to see problems outside of backup & restore code.

Overall changes look good though. Just trying to reconcile my understanding of the problem with the way it's described in the PR.

achow101 commented:
I think there's one more wrinkle to this issue: wallets outside of the wallets directory are probably stored where the wallet file is the name of the wallet, rather than the directory being the name, and containing a wallet.dat. An example of something similar to this is `test_direct_file` test case, although in that case the file is still in the wallets directory. In fact, I think this PR actually "breaks" that test case as the backup file is no longer named as we expect it (although the test does not check that, but it probably should).

I believe this PR would be incorrect for such wallets since it uses the parent directory's name for the backup, rather than the wallet file's name. For such wallets, the name of the parent directory would probably be something unrelated to the wallet file's name. If there are multiple wallet files in the same directory, this would also be problematic.

luke-jr reviewed (changes requested):
Won't this turn `/home/user/arbitrarywalletname.db` into `/home/user/user_%d.legacy.bak` ? Seems incorrect?

ryanofsky commented:
I was under impression that the only wallet paths that could be loaded are:

https://github.com/bitcoin/bitcoin/blob/06439a14c884d7f81f331646ad361e88b3037a51/src/wallet/wallet.cpp#L2974-L2977

So it would not be possible to load a bare db outside -walletdir. And if the migration code was backing up a bare db file inside -walletdir like `arbitrarywalletname.db` it would just be backed up as `arbitrarywalletname.db_1223.bak`

achow101 commented:
> So it would not be possible to load a bare db outside -walletdir.

Good point

> And if the migration code was backing up a bare db file inside -walletdir like `arbitrarywalletname.db` it would just be backed up as `arbitrarywalletname.db_1223.bak`

This PR currently breaks that. Although `test_direct_file` doesn't currently fail, adding a check for the wallet's name in the backup path does make it fail, e.g.

```diff
diff --git a/test/functional/wallet_migration.py b/test/functional/wallet_migration.py
index c467d6ad36d..90088db34e7 100755
--- a/test/functional/wallet_migration.py
+++ b/test/functional/wallet_migration.py
@@ -568,7 +568,8 @@ class WalletMigrationTest(BitcoinTestFramework):
         )
         assert (self.master_node.wallets_path / "plainfile").is_file()
 
-        self.master_node.migratewallet("plainfile")
+        migrate_res = self.master_node.migratewallet("plainfile")
+        assert "plainfile" in os.path.basename(migrate_res["backup_path"])
         wallet = self.master_node.get_wallet_rpc("plainfile")
         info = wallet.getwalletinfo()
         assert_equal(info["descriptors"], True)
```

ryanofsky commented:
Although `test_direct_file` doesn't currently fail, adding a check for the wallet's name in the backup path does make it fail, e.g.

That's a good point. Instead of backing up `plainfile` as `plainfile_123.legacy.bak` 47174476c04a7f1138fbd32eb0c9c38e85946393 backs it up as `wallets_123.legacy.bak` because it assumes the name of the parent directory is the name of the wallet. Following would seem to be a good fix:

```diff
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -4499,13 +4499,14 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
         return util::Error{_("Error: This wallet is already a descriptor wallet")};
     }
 
-    // Make a backup of the DB in the wallet's directory with a unique filename using the wallet
-    // name and current timestamp. If not the default wallet, the name of the folder containing
-    // wallet data is used as the wallet's name, since its full name may contain slashes.
-    fs::path legacy_wallet_dir = fs::absolute(fs::PathFromString(local_wallet->GetDatabase().Filename())).parent_path();
-    std::string legacy_wallet_dir_name = fs::PathToString(legacy_wallet_dir.filename());
-    fs::path backup_filename = fs::PathFromString(strprintf("%s_%d.legacy.bak", (wallet_name.empty() ? "default_wallet" : legacy_wallet_dir_name), GetTime()));
-    fs::path backup_path = legacy_wallet_dir / backup_filename;
+    // Make a backup of the DB in the wallet's directory with a unique filename
+    // using the wallet name and current timestamp. The backup filename is based
+    // on walletname but expanded to "default_wallet" if it is empty and only
+    // the final filename portion is used if contains slashes.
+    fs::path legacy_wallet_dir = fs::PathFromString(local_wallet->GetDatabase().Filename()).parent_path();
+    std::string backup_prefix = fs::PathToString(fs::PathFromString(wallet_name).filename());
+    fs::path backup_filename = fs::PathFromString(strprintf("%s_%d.legacy.bak", (backup_prefix.empty() ? "default_wallet" : backup_prefix), GetTime()));
+    fs::path backup_path = fsbridge::AbsPathJoin(legacy_wallet_dir, backup_filename);
     if (!local_wallet->BackupWallet(fs::PathToString(backup_path))) {
         if (was_loaded) {
             reload_wallet(local_wallet);
```

It also drops the incorrect and unnecessary fs::absolute call since CWD should be irrelevant here.

pablomartin4btc reviewed (commented):
tACK 6b4b7ac2113ed969bd31ce42190b96bd2e1b9c97 

I was able to reproduce the issue on `master`. As a detail, the CLI outputs "_Unable to make a backup of your wallet_" when the backup fails—might be worth mentioning that in the PR description.

This branch fixes the problem as expected (checked both relatives `.../../` and `a/b/c`).

Note that legacy wallet creation is no longer possible on `master`, so it might be helpful to update the reproduction steps. One option could be to use a previous release (e.g., v28.0 - `./releases/v28.0/bin/bitcoind`) via previously `./test/get_previous_releases.py -b`.

> https://github.com/bitcoin/bitcoin/commit/47174476c04a7f1138fbd32eb0c9c38e85946393 backs it up as wallets_123.legacy.bak

I've verified that the [above](https://github.com/bitcoin/bitcoin/pull/32273#issuecomment-2821309098) needs to be fixed for **plainfile** wallets migration.

davidgumberg committed:
test: wallet: Check direct file backup name.

This check ensures that when migrating a legacy wallet with a direct
filename, the backup file is named as expected.

Co-authored-by: Ava Chow <github@achow101.com>

davidgumberg commented on test/functional/wallet_migration.py:
```diff
@@ -925,6 +990,54 @@ def test_failed_migration_cleanup(self):
+            assert_equal(magic, BTREE_MAGIC)
+
+
     def test_blank(self):
```
Fixed, thanks.

davidgumberg committed:
wallet: migration: Make backup in walletdir

davidgumberg commented:
Rebased because of spurious CI failure fixed by #32859.

furszy commented on test/functional/wallet_migration.py:
```diff
@@ -148,7 +148,10 @@ def test_basic(self):
         # Note: migration could take a while.
-        _, basic0 = self.migrate_and_get_rpc("basic0")
+        res0, basic0 = self.migrate_and_get_rpc("basic0")
+        backup_path = self.master_node.wallets_path / f"basic0_{int(time.time())}.legacy.bak"
+        assert backup_path.exists()
+        assert_equal(str(backup_path), res0['backup_path'])
```
We are already doing the backup file existence check at the end of `migrate_and_get_rpc`, what if you include this extra check there too?

davidgumberg replied:
Thanks for this suggestion, I've implemented this approach in https://github.com/bitcoin/bitcoin/pull/32273/commits/55e60bd9c76ea74a108cd5c7275e26554b822b4c, and I've reused `migrate_and_get_rpc()` in the other tests added here (except [`test_failed_migration_cleanup_relative_path()`](https://github.com/bitcoin/bitcoin/blob/6b2e3fbf64b19b6065768ef85155f986859fcfb9/test/functional/wallet_migration.py#L1038) where it's not feasible)

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/32273#issuecomment-3029455047

Sorry I didn't post my complete suggestion before. Here is the change I would suggest based on the current PR (2e8a21a43a3c321c9f3a8c75024d036165658225):

<details><summary>diff</summary>
<p>

```diff
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -4253,30 +4253,16 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
 
     // Make a backup of the DB in the wallet's directory with a unique filename
     // using the wallet name and current timestamp. The backup filename is based
-    // on the name of the parent directory containing the wallet data in most
-    // cases, but in the case where the wallet name is a path to a data file,
-    // the name of the data file is used, and in the case where the wallet name
-    // is blank, "default_wallet" is used.
-    auto legacy_wallet_path = GetWalletPath(wallet_name);
-    if (!legacy_wallet_path) {
-        return util::Error{util::ErrorString(legacy_wallet_path)};
-    }
-
-    // Backwards-compatibility logic for migrating wallets whose names are
-    // paths to data files directly rather than directories.
-    bool is_direct_file = fs::symlink_status(*legacy_wallet_path).type() == fs::file_type::regular;
-    std::string backup_prefix;
-    if (wallet_name.empty()) {
-        backup_prefix = "default_wallet";
-    } else if (is_direct_file) {
-        // We use filename() to strip slashes in the direct file name.
-        backup_prefix = fs::PathToString(fs::PathFromString(wallet_name).filename());
-    } else {
-        // The canonical form resolves relative specifiers in the
-        // wallet's path, useful for the backup prefix.
-        auto legacy_wallet_dir = fs::weakly_canonical(*legacy_wallet_path);
-        backup_prefix = fs::PathToString(legacy_wallet_dir.filename());
-    }
+    // on walletname but expanded to "default_wallet" if it is empty and only
+    // the final filename portion is used if contains slashes.
+    std::string backup_prefix = wallet_name.empty() ? "default_wallet" : [&] {
+        // Normalize any .. components
+        fs::path p{fs::PathFromString(wallet_name).lexically_normal()};
+        // Drop trailing slash if any.
+        if (p.filename().empty()) p = p.parent_path();
+        // Return last path component
+        return fs::PathToString(p.filename());
+    }();
     fs::path backup_filename = fs::PathFromString(strprintf("%s_%d.legacy.bak", backup_prefix, GetTime()));
     fs::path backup_path = fsbridge::AbsPathJoin(GetWalletDir(), backup_filename);
     if (!local_wallet->BackupWallet(fs::PathToString(backup_path))) {
```
</p>
</details>

There isn't a reason I can see to access the filesystem just to convert a path into a backup prefix. IMO doing this makes the code unnecessarily fragile and complicated. I also don't think it's good to add a call to the `GetWalletPath()` function which is complicated enough to understand in the context of wallet loading, and shouldn't be responsible for supporting wallet migration code as well.

I don't think belt and suspenders checks are a great idea in general, but even when they are appropriate I think they need to be labeled with "// This condition is never expected to trigger" or similar comments to avoid confusing people reading the code and trying understand why the checks are there.

> Agreed, I've added [75df713](https://github.com/bitcoin/bitcoin/commit/75df71398dc4feefaa6f6ee611e41c5daa8ebc2b) which moves the backup from inside the wallet directory to being in the node's `-walletdir`.

Nice! I actually misinterpreted your commit and didn't realize storing backups inside wallet directories was actually previous behavior, but I think changing this does make sense and it is good to be able to remove the extra copy_file calls in the code.

> Unfortunately, I think knowing the name of the wallet's parent folder requires filesystem access when arbitrary relative paths are permitted. `lexically_normal()` won't work in test cases like [8e20619](https://github.com/bitcoin/bitcoin/commit/8e20619341b465e30bdbfc22e6be515fb2ae2e6f) where the wallet is specified by a relative path that ascends more than it descends, or [e.g. the wallet named `../`.](https://godbolt.org/z/eTnhqe8de)

I don't think this is true. If a wallet ascends more than it descends like `../../../../mywallet`, the suggested change would use "mywallet" as the prefix. If the path contains `.` components or internal `..`, `lexically_normal` will strip them out. If the path contains trailing slashes, the suggested code strips them out to. If the path is literally just `..` or `../..` or similar even that is fine, producing backups files like `.._1234.legacy.bak`

> This code could just as well be: [...]

Yes, and I do think that change would be a clear simplification with no downsides. But I think it would be good to simplify even further and not resolve symlinks, not check to see if wallets are files or directories. None of this seems very helpful for choosing a good backup prefix.

davidgumberg committed:
test: Migration of a wallet ending in `/`

davidgumberg committed:
test: Migration of a wallet ending in `../`

davidgumberg commented on test/functional/wallet_migration.py:
```diff
@@ -551,6 +551,38 @@ def test_unloaded_by_path(self):
+        wallet_path = f"{self.old_node.wallets_path}/{wallet_name}"
+        migrate_res = self.master_node.migratewallet(wallet_path)
+        expected_backup_path = self.master_node.wallets_path / f"{final_dir}_{int(time.time())}.legacy.bak"
+        os.path.exists(expected_backup_path)
```
As suggested by @furszy [above](https://github.com/bitcoin/bitcoin/pull/32273#discussion_r2181204274) I moved this logic to `migrate_and_get_rpc()` in https://github.com/bitcoin/bitcoin/commit/55e60bd9c76ea74a108cd5c7275e26554b822b4c:

https://github.com/bitcoin/bitcoin/blob/6b2e3fbf64b19b6065768ef85155f986859fcfb9/test/functional/wallet_migration.py#L125-L147

And I've reused `migrate_and_get_rpc()` in the tests added in this PR.

davidgumberg commented:
> I don't think belt and suspenders checks are a great idea in general, but even when they are appropriate I think they need to be labeled with "// This condition is never expected to trigger" or similar comments to avoid confusing people reading the code and trying understand why the checks are there.

OK, in lieu of an assert/assume, I've gone to the simplified form suggested, and I've also used the lambda+ternary over if+else form as you suggested. I've also taken advantage of the fact that direct files will never have slashes in their names, so we can use the same logic for direct files and paths in picking a prefix:

https://github.com/bitcoin/bitcoin/blob/9fa5480fc4c46fa11345c45fbe4dd21c127e3e05/src/wallet/wallet.cpp#L4260-L4267

But, I have retained the use of `fs::weakly_canonical`, since:

> If the path is literally just .. or ../.. or similar even that is fine, producing backups files like .._1234.legacy.bak

While better than master, I don't think this is the right behavior, and I'm not sure I follow why it's fragile or complicated for us to ask the filesystem to resolve the canonical names of relative paths, and it seems to me like the only way to preserve the most sensible behavior, of naming the wallet backup after the name of the wallet directory it comes from.

achow101 commented on test/functional/wallet_migration.py:
```diff
@@ -997,6 +997,53 @@ def test_failed_migration_cleanup(self):
+
+        assert all(wallet not in self.master_node.listwallets() for wallet in [f"{wallet_name}", f"{wallet_name}_watchonly", f"{wallet_name}_solvables"])
+
+        assert not (self.master_node.wallets_path / "failed_watchonly").exists()
```
In e7fec273593d825e0e7dbec46a026aacc58c479f "test: Migration fail recovery w/ `../` in path"

I think `failed_watchonly` is the wrong wallet name. Probably best to do `f"{wallet_name}_watchonly"`

davidgumberg replied:
Thanks for catching this, I've applied your fix in f67d6fe811

achow101 commented on test/functional/wallet_migration.py:
```diff
@@ -618,6 +618,26 @@ def test_wallet_with_relative_path(self):
+        self.generate(self.master_node, 1)
+        bals = wallet.getbalances()
+
+        _, wallet = self.migrate_and_get_rpc(wallet_name)
```
In 146b133697f07c8dda004d60087505b2e1ec14b9 "test: Migration of a wallet ending in `/`"

Could add a check for the expected backup path.

davidgumberg replied:
I add this check to `migrate_and_get_rpc()` in 55e60bd9c76:

https://github.com/bitcoin/bitcoin/blob/9fa5480fc4c46fa11345c45fbe4dd21c127e3e05/test/functional/wallet_migration.py#L138-L146

achow101 commented on test/functional/wallet_migration.py:
```diff
@@ -638,6 +638,24 @@ def test_wallet_with_path_ending_in_slash(self):
+
+        _, wallet = self.migrate_and_get_rpc(wallet_ending_in_relative)
+
+        assert wallet.gettransaction(txid)
```
In 9fa5480fc4c46fa11345c45fbe4dd21c127e3e05 "test: Migration of a wallet ending in `../`"

Could add a check for the expected backup path.

davidgumberg replied:
See above: https://github.com/bitcoin/bitcoin/pull/32273#discussion_r2183974514

pablomartin4btc commented on test/functional/wallet_migration.py:
```diff
@@ -585,7 +585,8 @@ def test_direct_file(self):
         assert (self.master_node.wallets_path / "plainfile").is_file()
 
-        self.master_node.migratewallet("plainfile")
+        migrate_res = self.master_node.migratewallet("plainfile")
+        assert_equal(f"plainfile_{int(time.time())}.legacy.bak", os.path.basename(migrate_res["backup_path"]))
```
Maybe you can use `mocktime` here for consistency?

```diff --git a/test/functional/wallet_migration.py b/test/functional/wallet_migration.py
--- a/test/functional/wallet_migration.py
+++ b/test/functional/wallet_migration.py
@@ -585,8 +585,10 @@ class WalletMigrationTest(BitcoinTestFramework):
         )
         assert (self.master_node.wallets_path / "plainfile").is_file()
 
+        mocked_time = int(time.time())
+        self.master_node.setmocktime(mocked_time)
         migrate_res = self.master_node.migratewallet("plainfile")
-        assert_equal(f"plainfile_{int(time.time())}.legacy.bak", os.path.basename(migrate_res["backup_path"]))
+        assert_equal(f"plainfile_{mocked_time}.legacy.bak", os.path.basename(migrate_res["backup_path"]))

```

davidgumberg replied:
Thank you, fixed!

pablomartin4btc reviewed (commented):
tACK 76fe0e59ec4a5b0c5b18f46bfcbbf99628d74d77

I've re-[tested](https://github.com/bitcoin/bitcoin/pull/32273#pullrequestreview-2964023741) the relative path cases and the plainfile [issue](https://github.com/bitcoin/bitcoin/pull/32273#issuecomment-2819784017) has been fixed.

Regarding the usage of `fs::weakly_canonical` vs `lexically_normal()`, I agree with @ryanofsky, I think that resource cost-wise, to "access" the filesystem in order to validate the name that we'll use for the backup file that it'll be written and validated then (its existence/ permission and so on). If there's something that `lexically_normal()` can't do right comparing to what `fs::weakly_canonical` does, I think we should go around that logic (if possible).

e.g. `fs::path("/foo/./bar/../baz").lexically_normal()  → "/foo/baz"` (even if /foo/baz doesn't exist — this will still work).

In any case, are we logging the before & after applying this parsing (using `fs::weakly_canonical` or `lexically_normal()`)? Also perhaps we should document some examples somewhere so the users can find out why something they tried didn't meet their expectations (was accidentally stored somewhere else or with a different name).

achow101 reviewed (commented):
ACK 76fe0e59ec4a5b0c5b18f46bfcbbf99628d74d77

ryanofsky commented on test/functional/wallet_migration.py:
```diff
@@ -637,6 +637,24 @@ def test_wallet_with_path_ending_in_slash(self):
 
+    def test_wallet_with_path_ending_in_relative_specifier(self):
+        self.log.info("Test migrating a wallet with a name/path ending in a relative specifier, '..'")
+        wallet_ending_in_relative = "path/that/ends/in/.."
```
In commit "test: Migration of a wallet ending in `../`" (76fe0e59ec4a5b0c5b18f46bfcbbf99628d74d77)

Note: there's a minor typo in the commit message. `../` should just be `..` since the path doesn't end in a slash.

Also this commit seems very similar to previous commit. Could be nice to combine these into a single function like:

```py
self.test_wallet_with_path("path/to/mywallet/")
self.test_wallet_with_path("path/that/ends/in/..")
```

With a `fail_migration: bool` option, this could be extended to dedup `test_wallet_with_relative_path` and `test_failed_migration_cleanup_relative_path` cases as well.

Seem ok to have duplicate test code for now though.

ryanofsky reviewed (approved):
Code review ACK 76fe0e59ec4a5b0c5b18f46bfcbbf99628d74d77. Nice changes that (1) fix potential errors when names of wallets being migrated contain slashes, and (2) store migration backups in the top-level `-walletdir` instead of in individual wallet subdirectories.

I left various comments below but none are important. I re-reviewed this PR from the beginning since there were a number of changes since my last review.

achow101 commented:
I'm going to go ahead and merge this now as @davidgumberg is out for the next couple weeks. Remaining comments can be addressed in a followup.

kevkevinpal commented on test/functional/wallet_migration.py:
```diff
@@ -560,6 +563,60 @@ def test_unloaded_by_path(self):
+        # migratewallet uses current time in naming the backup file, set a mock time
+        # to check that this works correctly.
+        curr_time = int(time.time())
+        self.master_node.setmocktime(curr_time)
```
looks like this was fixed in https://github.com/bitcoin/bitcoin/pull/33104

kevkevinpal commented on src/wallet/wallet.cpp:
```diff
@@ -4229,8 +4229,19 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
+    // is blank, "default_wallet" is used.
+    const std::string backup_prefix = wallet_name.empty() ? "default_wallet" : [&] {
+        // fs::weakly_canonical resolves relative specifiers and remove trailing slashes.
+        const auto legacy_wallet_path = fs::weakly_canonical(GetWalletDir() / fs::PathFromString(wallet_name));
```
addressing this in https://github.com/bitcoin/bitcoin/pull/33122

ryanofsky replied:
It was discussed, see https://github.com/bitcoin/bitcoin/pull/32273#pullrequestreview-2983599593:

> I don't think this is true. If a wallet ascends more than it descends like `../../../../mywallet`, the suggested change would use "mywallet" as the prefix. If the path contains `.` components or internal `..`, `lexically_normal` will strip them out. If the path contains trailing slashes, the suggested code strips them out to. If the path is literally just `..` or `../..` or similar even that is fine, producing backups files like `.._1234.legacy.bak`

More generally, putting this specific suggestion aside, hopefully it is clear that turning a wallet name into a safe backup filename just requires string manipulation, and does not require accessing the filesystem, resolving symlinks, or calling GetWalletDir().
